ssl/quic/quic_rx_frame.c

```diff
@@ -0,0 +1,391 @@
+/*
+ * Copyright 2022 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <internal/packet.h>
+#include <internal/nelem.h>
+#include <internal/quic_wire.h>
+/* ASSUMPTION: this is the OSSL_QRL header file */
+#include <internal/quic_record_layer.h>
+
+static int depack_collect_ack_data(PACKET *pkt, OSSL_ACKM_RX_PKT *ackm_data)
+{
+    /* Initialize by pretending the packet isn't ACK eliciting */
+    ackm_data->is_ack_eliciting = 0;
+
+    while (PACKET_remaining(&pkt) > 0) {
+        if (!ossl_quic_wire_peek_frame_header(&pkt, &frame_type))
+            return 0;
+
+        switch (frame_type) {
+        /* Ignore all frames that aren't ACK eliciting */
+        case OSSL_QUIC_FRAME_TYPE_PING:
+        case OSSL_QUIC_FRAME_TYPE_ACK_WITHOUT_ECN:
+        case OSSL_QUIC_FRAME_TYPE_ACK_WITH_ECN:
+        case OSSL_QUIC_FRAME_TYPE_CONN_CLOSE_TRANSPORT:
+        case OSSL_QUIC_FRAME_TYPE_CONN_CLOSE_APP:
+            break;
+        default:
+            /* This frame made the packet ACK eliciting */
+            ackm_data->is_ack_eliciting = 0;
+            break;
+        }
+    }
+}
+
+static int depack_process_frames(PACKET *pkt, int packet_space,
+                                 OSSL_TIME received)
+{
+    while (PACKET_remaining(&pkt) > 0) {
+        if (!ossl_quic_wire_peek_frame_header(&pkt, &frame_type))
+            return 0;
+
+        switch (frame_type) {
+        case OSSL_QUIC_FRAME_TYPE_PADDING:
+        case OSSL_QUIC_FRAME_TYPE_PING:
+            /*
+             * We ignore this.
+             * For the PING frame, eliciting an ACK is enough.
+             * Since the PING frame is only a header, we simply
+             * skip it.
+             */
+            ossl_quic_wire_skip_frame_header(&pkt, &frame_type);
+            break;
+
+        case OSSL_QUIC_FRAME_TYPE_ACK_WITHOUT_ECN:
+        case OSSL_QUIC_FRAME_TYPE_ACK_WITH_ECN: {
+            OSSL_ACKM_ACK_RANGE ranges[20];
+            size_t total_ranges;
+            OSSL_ACKM_ACK ack;
+
+            ack.ack_ranges = ranges;
+            ack.num_ack_ranges = OSSL_NELEM(ranges);
+            /* QUESTION: where do I get the ack delay exponent? */
+            if (!ossl_quic_wire_decode_frame_ack(pkt, ack_delay_exponent,
```

Comment:
This function is designed to be able to pass a NULL `ack` value initially and return the total_ranges so that you can then allocate sufficient space for them.

Reply:
Ah, that wasn't the case last time I looked...  thanks

Reply:
Fixed